src/Lykke.Job.TransactionHandler/Sagas/EthereumCoreSaga.cs

```diff
@@ -0,0 +1,136 @@
+using Lykke.Job.TransactionHandler.Core.Domain.PaymentSystems;
+using Lykke.Job.TransactionHandler.Core.Services.BitCoin;
+using Lykke.Job.TransactionHandler.Events;
+using Lykke.Job.TransactionHandler.Events.EthereumCore;
+using Lykke.Service.Assets.Client;
+
+namespace Lykke.Job.TransactionHandler.Sagas
+{
+    public class EthereumCoreSaga
+    {
+        private readonly ILog _log;
+        private readonly IEthereumCashinAggregateRepository _ethereumCashinAggregateRepository;
+
+        public EthereumCoreSaga(
+            [NotNull] ILog log,
+            [NotNull] IEthereumCashinAggregateRepository ethereumCashinAggregateRepository)
+        {
+            _log = log?.CreateComponentScope(nameof(EthereumCoreSaga)) ?? throw new ArgumentNullException(nameof(log));
+            _ethereumCashinAggregateRepository = ethereumCashinAggregateRepository;
+        }
+
+        private async Task Handle(CashinDetectedEvent evt, ICommandSender sender)
+        {
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Start();
+
+                _log.WriteInfo(evt.TransactionHash, evt, "Eth Cashin startetd");
+
+                var aggregate = await _ethereumCashinAggregateRepository.GetOrAddAsync(evt.TransactionHash, () =>
+                                    new EthereumCashinAggregate(evt.TransactionHash, evt.ClientId, evt.AssetId,
+                                    evt.ClientAddress, evt.Amount, evt.CreatePendingActions));
+
+                if (aggregate.State == EthereumCashinState.CashinStarted)
+                {
+                    sender.SendCommand(new EnrollEthCashinToMatchingEngineCommand()
+                    {
+                        TransactionHash = evt.TransactionHash,
+                        Amount = evt.Amount,
+                        AssetId = evt.AssetId,
+                        ClientAddress = evt.ClientAddress,
+                        ClientId = evt.ClientId,
+                        CreatePendingActions = evt.CreatePendingActions,
+                        CashinOperationId = aggregate.CashinOperationId
+                    }, BoundedContexts.EthereumCommands);
+                }
+            }
+            catch (Exception e)
+            {
+                _log.WriteError(evt.TransactionHash, evt, e);
+                throw;
+            }
+            finally
+            {
+                _log.WriteInfo(evt.TransactionHash, evt, $"Eth Cashin start completed in {sw.ElapsedMilliseconds}");
+                sw.Stop();
+            }
+        }
+
+        private async Task Handle(EthCashinEnrolledToMatchingEngineEvent evt, ICommandSender sender)
+        {
+            Stopwatch sw = new Stopwatch();
+            try
+            {
+                sw.Start();
+                _log.WriteInfo(evt.TransactionHash, evt, "Cashin Enrolled To ME");
+
+                var aggregate = await _ethereumCashinAggregateRepository.TryGetAsync(evt.TransactionHash);
```

Comment:
Why you use TryGetAsync, but not GetAsync? aggregate can be missed here? Not looks so, but if it can, then it should be tested for null before accessing

Reply:
fixed